src/qtoolkit/io/sge.py

```diff
@@ -0,0 +1,526 @@
+            )
+        match = re.search(r'Your job (\d+) \(".*?"\) has been submitted', stdout)
+        if not match:
+            raise OutputParsingError("Failed to parse job ID from stdout")
+        job_id = match.group(1)
+        status = (
+            SubmissionStatus("SUCCESSFUL")
+            if job_id
+            else SubmissionStatus("JOB_ID_UNKNOWN")
+        )
+        return SubmissionResult(
+            job_id=job_id,
+            exit_code=exit_code,
+            stdout=stdout,
+            stderr=stderr,
+            status=status,
+        )
+
+    def parse_cancel_output(self, exit_code, stdout, stderr) -> CancelResult:
+        """Parse the output of the qdel command."""
+        if isinstance(stdout, bytes):
+            stdout = stdout.decode()
+        if isinstance(stderr, bytes):
+            stderr = stderr.decode()
+        if exit_code != 0:
+            return CancelResult(
+                exit_code=exit_code,
+                stdout=stdout,
+                stderr=stderr,
+                status=CancelStatus("FAILED"),
+            )
+        match = re.search(r"qdel: job (\d+) deleted", stderr)
+        if not match:
+            raise OutputParsingError("Failed to parse job ID from stdout")
+        job_id = match.group(1)
+        status = CancelStatus("SUCCESSFUL")
+        return CancelResult(
+            job_id=job_id,
+            exit_code=exit_code,
+            stdout=stdout,
+            stderr=stderr,
+            status=status,
+        )
+
+    def _get_job_cmd(self, job_id: str):
+        cmd = f"qstat -j {job_id}"
+        return cmd
+
+    def parse_job_output(self, exit_code, stdout, stderr) -> QJob | None:  # aiida style
+        if exit_code != 0:
+            raise OutputParsingError(f"Error in job output parsing: {stderr}")
+        if isinstance(stdout, bytes):
+            stdout = stdout.decode()
+        if isinstance(stderr, bytes):
+            stderr = stderr.decode()
+
+        # Check for specific error messages in stderr or stdout
+        error_patterns = [
+            re.compile(
+                r"Primary job\s+terminated normally, but\s+(\d+)\s+process returned a non-zero exit code",
+                re.IGNORECASE,
+            ),
+            re.compile(
+                r"mpiexec detected that one or more processes exited with non-zero status",
+                re.IGNORECASE,
+            ),
+            re.compile(r"An error occurred in MPI_Allreduce", re.IGNORECASE),
+            re.compile(
+                r"Error: mca_pml_ucx_send_nbr failed: -25, Connection reset by remote peer",
+                re.IGNORECASE,
+            ),
+            re.compile(r"mpi_errors_are_fatal", re.IGNORECASE),
+        ]
+
+        for pattern in error_patterns:
+            if pattern.search(stderr) or pattern.search(stdout):
+                raise OutputParsingError(
+                    "Job terminated due to a non-zero exit code from one or more processes or MPI errors"
```

Comment:
Perhaps we can add the full stderr here too?

```suggestion
                    "Job terminated due to a non-zero exit code from one or more processes or MPI errors: {stderr}"
```